pkg/pod/pod.go

```diff
@@ -98,6 +105,52 @@ func (c *Client) GetPods() (pods []*v1.Pod, err error) {
 	return listPods, nil
 }
 
+// ListPods returns list of all pods
+func (c *Client) ListPods() (pods []*v1.Pod, err error) {
+	var resList []*v1.Pod
+	listPods := c.PodWatcher.Informer().GetStore().List()
+
+	for _, pod := range listPods {
+		v1Pod, ok := pod.(*v1.Pod)
+		if !ok {
+			err := fmt.Errorf("could not cast %T to v1.Pod", pod)
+			klog.Error(err)
+			return nil, err
+		}
+
+		resList = append(resList, v1Pod)
+		klog.V(6).Infof("Appending pod ID: %s/%s to list.", v1Pod.UID, v1Pod.Name)
+	}
+
+	return resList, nil
+}
+
+// GetPod returns the pod object
+func GetPod(obj interface{}, i informersv1.PodInformer) (pod *v1.Pod, err error) {
+	currentPod, exists, err := i.Informer().GetStore().Get(obj)
+	if !exists || err != nil {
+		err := fmt.Errorf("Could not get Pod %s", obj.(*v1.Pod))
+		klog.Error(err)
+		return nil, err
+	}
+
+	pod = currentPod.(*v1.Pod)
+	for pod.Status.PodIP == "" || pod.Spec.NodeName == "" {
+		fmt.Printf("Sleep 1 second to wait for pod ip and node name\n")
+		time.Sleep(1 * time.Second)
```

Comment:
Seeing such sleeps in the code is usually an indicator of some timing issue in the code. Why do you need to sleep for 1 second?

Reply:
When AKS starts to deploy a pod, the pod creation event will be fired. It always happen before the pod is completed created, which means when we detect the pod creation event, the pod IP is not assigned to the pod yet. So add this check here. I tested a bunch of pods and most of them needs wait 4~5 seconds to get the pod id assigned.

Reply:
I'd suggest giving this an upper limit, so we're not blocked forever in the start of an error condition.